katdal/categorical.py

```diff
@@ -17,107 +17,93 @@
+               (self.__class__.__name__, self.unwrapped, id(self))
 
     def __str__(self):
         """Longer human-friendly string representation of wrapped object."""
-        return str(self.unwrap())
+        return str(self.unwrapped)
 
     def __eq__(self, other):
         """Equality comparison operator."""
-        if not isinstance(other, HashableValueWrapper):
-            other = HashableValueWrapper.wrap(other)
-        return self.hashable_value == other.hashable_value
+        if isinstance(other, ComparableArrayWrapper):
+            other = other.unwrapped
+        if isinstance(self.unwrapped, np.ndarray) or isinstance(other, np.ndarray):
+            return np.array_equal(self.unwrapped, other)
+        else:
+            return self.unwrapped == other
 
     def __ne__(self, other):
         """Inequality comparison operator."""
         return not self == other
 
     def __lt__(self, other):
         """Less-than comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped < other
 
     def __gt__(self, other):
         """Greather-than comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped > other
 
     def __le__(self, other):
         """Less-than-or-equal comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped <= other
 
     def __ge__(self, other):
         """Greater-than-or-equal comparison operator."""
-        raise TypeError("HashableValueWrapper objects are considered unorderable")
+        return self.unwrapped >= other
 
-    def __hash__(self):
-        """Return hash value of wrapped object (guaranteed to be hashable)."""
-        return hash(self.hashable_value)
-
-    @staticmethod
-    def wrap(value):
-        """Wrap generic object to become hashable (pickle implementation)."""
-        return HashableValueWrapper(pickle.dumps(value))
-
-    def unwrap(self):
-        """Extract original object from wrapper (pickle implementation)."""
-        return pickle.loads(self.hashable_value)
+    # Force class to be unhashable to allow value comparison via __eq__ if possible
+    __hash__ = None
 
 
 def unique_in_order(elements, unwrap=False, return_inverse=False):
     """Extract unique elements from `elements` while preserving original order.
 
     Parameters
     ----------
-    elements : sequence of hashables
-        Sequence of hashable objects (i.e. objects supporting __hash__ operator)
+    elements : sequence of hashables or comparables
+        Sequence of hashable or comparable objects (i.e. objects supporting
+        either __hash__ or __eq__ operator)
     unwrap : {False, True}, optional
-        If True, unwrap any objects wrapped in a :class:`HashableValueWrapper`
+        If True, unwrap any objects wrapped in a :class:`ComparableArrayWrapper`
```

Comment:
I thought the same thing and it's already on the chopping block. The ```unwrap=True``` setting only occurs in two special places were it can easily be done outside the function.